src/shared/nuclearnet/tests/nuclearnet_proxy_parser.tests.ts

```diff
@@ -0,0 +1,170 @@
+    packets.forEach(p => e.encode(p, (chunks: any[]) => wire.push(...chunks)))
+
+    // Decode all the packets
+    const decoded: any[] = []
+    d.on('decoded', (packet: any) => decoded.push(packet))
+    wire.forEach(p => d.add(p))
+
+    expect(decoded).toEqual(packets)
+  })
+
+  it('Roundtrips NUClearNet informational packets', () => {
+
+    const packets: Packet[] = [
+      {
+        nsp: '/',
+        type: TYPES.EVENT,
+        data: ['nuclear_join', {
+          name: 'itsnotme',
+          address: '127.0.0.1',
+          port: 40000 + Math.floor(Math.random() * 10000),
+        }],
+        id: Math.floor(Math.random() * 1000),
+      },
+      {
+        nsp: '/',
+        type: TYPES.EVENT,
+        data: ['nuclear_leave', {
+          name: 'itsyou',
+          address: '192.168.10.11',
+          port: 40000 + Math.floor(Math.random() * 10000),
+        }],
+        id: Math.floor(Math.random() * 1000),
+      },
+      {
+        nsp: '/',
+        type: TYPES.EVENT,
+        data: ['nuclear_connect', {}],
+        id: Math.floor(Math.random() * 1000),
+      },
+      {
+        nsp: '/',
+        type: TYPES.EVENT,
+        data: ['nuclear_disconnect'],
+        id: Math.floor(Math.random() * 1000),
+      },
+      {
+        nsp: '/',
+        type: TYPES.EVENT,
+        data: ['listen', 'message.input.Image', Math.floor(Math.random() * 1000)],
+        id: Math.floor(Math.random() * 1000),
+      },
+      {
+        nsp: '/',
+        type: TYPES.EVENT,
+        data: ['unlisten', Math.floor(Math.random() * 1000)],
+        id: Math.floor(Math.random() * 1000),
+      },
+    ]
+
+    // Encode all the packets
+    const wire: any[] = []
+    packets.forEach(p => e.encode(p, (chunks: any[]) => wire.push(...chunks)))
+
+    // Decode all the packets
+    const decoded: any[] = []
+    d.on('decoded', (packet: any) => decoded.push(packet))
+    wire.forEach(p => d.add(p))
+
+    expect(decoded).toEqual(packets)
+  })
+
+  it('Roundtrips NUClearNet packets with binary data in them', () => {
+
+    // Encode an image protocol buffer as an example
+    //
+    const nuclearPackets: NUClearNetPacket[] = [
+      {
+        peer: {
+          name: 'itshard',
+          address: '192.168.2.11',
+          port: 40000 + Math.floor(Math.random() * 10000),
+        },
+        hash: new Buffer(8),
+        payload: new Buffer(1000),
+        reliable: true,
+      },
+      {
+        peer: {
+          name: 'tothinkupnames',
+          address: '192.168.2.11',
+          port: 40000 + Math.floor(Math.random() * 10000),
+        },
+        hash: new Buffer(8),
+        payload: new Buffer(1000),
+        reliable: true,
+      },
+    ]
+
+    const packets: Packet[] = nuclearPackets.map((p): EventPacket => (
+    {
+      nsp: '/',
+      type: TYPES.EVENT,
+      data: ['message.input.Image', p],
+      id: Math.floor(Math.random() * 1000),
+    }
+    ))
+
+    // Encode all the packets
+    const wire: any[] = []
```

Comment:
Why `any`? Isn't this type `Packet[]`?

Reply:
No, this is the "on the wire' representation after the packets have been encoded.
According to the specs, they can be anything, although in this case for this parser they will be `string|Buffer` so to match what socket.io expects you to put on the wire I have left it as any.